O teste $\chi^2$ (lê-se *qui-quadrado*) é um procedimento projetado para testar a existência de relações significativas entre duas variáveis discretas.

# Passos do procedimento

1. Declarar as hipóteses nula e alternativa
2. Montar a tabela de contingência
3. Computar a estatística de teste
4. Calcular o valor $p$ correspondente
5. Tomar uma decisão e interpretar o resultado


# Tabela de contingência

Para aplicação do teste $\chi^2$, as informações coletadas (a partir da amostra) acerca das variáveis devem estar ​​organizadas em uma *tabela de contingência*.

Uma tabela de contingência (também conhecida como tabulação cruzada ou tabela de referência cruzada) é um tipo de tabela em um formato de matriz que exibe a distribuição de frequência das diversas combinações de valores das variáveis. Tabelas de contingência são muito utilizados em pesquisa de opinião, inteligência de negócios, engenharia e pesquisa científica. Elas fornecem uma imagem básica da inter-relação entre duas variáveis ​​e podem ajudar a encontrar interações entre elas.

Como exemplo, considere duas variáveis, ambas discretas: *sexo* (com valores em {'masculino' ou 'feminino'}) e *animal de estimação* (com valores em {'gato' ou 'cachorro'}). Além disso, suponha que 100 indivíduos sejam amostrados aleatoriamente em uma população muito grande como parte de um estudo de diferenças de sexo e preferência por animal de estimação. Uma tabela de contingência pode ser criada para exibir as diversas quantidades de indivíduos na amostra: quantos são homens e preferem gatos, quantos homens preferem gatos, quantas mulheres preferem cachorros e quantas mulheres preferem gatos. Essa tabela de contingência é exibida abaixo ([fonte](https://math.stackexchange.com/questions/1233587/checking-dependency-of-variables-in-a-contingency-table)).

![link text](https://i.stack.imgur.com/8YuSA.jpg)

Ainda com relação ao exemplo acima, repare que

- a coluna mais à direita apresenta os totais por linha,
- a linha mais abaixo exibe os totais por coluna,
- a célula inferior direita o total geral de indivíduos da amostra.

As informações dos totais por linha e por coluna são opcionais na representação de uma tabela de contingência. Além disso, tabelas de contingência não precisam envolver apenas variáveis binárias. Contanto que as variáveis sejam discretas (e com uma quantidade de valores dentro do aceitável), uma tabela de contingência pode ser usada para exibir sumários acerca delas, independentemente da *aridade* (i.e., da quantidade de valores) dessas variáveis. No exemplo a seguir ([fonte](http://www.sthda.com/sthda/RDoc/data/housetasks.txt
)), as variáveis apresentam quatro e treze valores, respectivamente. Note também que não são fornecidos os totais nas margens da tabela.

![](https://i0.wp.com/www.sthda.com/sthda/RDoc/images/ca-housetasks.png?w=578)



# Independência estatística

A independência estatística é um conceito importante na Teoria das Probabilidades. Considere dois eventos $A$ e $B$. Se o conhecimento da ocorrência do evento $A$ não muda a probabilidade de ocorrência do evento $B$, então escrevemos:

$$
\Pr(B|A) = \Pr(B)
$$

Sob a condição acima, a probabilidade conjunta dos eventos $A$ e $B$ é igual ao produto das probabilidades dos eventos $A$ e $B$: $\Pr(A,B) = \Pr(A) \times \Pr(B)$. Esses eventos são então considerados **estatisticamente independentes**.

A independência estatística não deve ser confundidade com outra propriedade que pode existir entre eventos, a exclusividade mútua. Se a probabilidade conjunta de dois eventos $A$ e $B$ for zero, esses eventos são ditos mutuamente exclusivos ou disjuntos. O seguinte destaca as condições sob as quais dois eventos são estatisticamente independentes e mutuamente exclusivos:

- $\Pr(A,B) = \Pr(A) \times \Pr(B) \iff \text{eventos independentes}$
- $\Pr(A,B) = 0 \iff \text{eventos mutuamente exclusivos}$

No contexto da Inferência Estatística, o conceito de evento está relacionado aos valores que uma variável aleatória assume. Além disso, o conceito de independência estatística entre duas variáveis aleatórias pode ser usado como um indicador de presença ou ausência de dependência entre elas. Em particular, se duas variáveis são independentes, as distribuições percentuais na tabela de contingência são idênticas.

Como exemplo, considere novamente a tabela de contingência usada em exemplo anterior e reapresentada abaixo. Nesse contexto, há duas variáveis aleatórias binárias: $X$ = sexo e $Y$ = animal de estimação preferido. Os dados nessa tabela parecem indicar que as variáveis $X$ e $Y$ não são estatisticamente independentes. Para ententer porque, repare que a maioria das mulheres prefere gatos (39 contra 9), enquanto que a maioria dos homens prefere cachorros (42 contra 10). Se essas variáveis são efetivamente independentes, deveríamos esperar encontrar preferências aproximadamente iguais por gatos e cachorros, tanto no grupo de mulheres quanto no de homens.

![link text](https://i.stack.imgur.com/8YuSA.jpg)

Visto que os dados apresentados na tabela acima são correspondentes a uma amostra da população, surge a seguinte questão:

> O enviesamento observado é obra do acaso (e portanto não reflete a distribuição da população), ou é consequência de as variáveis serem estatisticamente dependentes?

A questão acima pode ser investigada por meio do teste $\chi^2$.

# Declaração das hipóteses

No teste $\chi^2$, as hipóteses devem ser declaradas da seguinte forma:

- A hipótese nula ($H_0$) deve afirmar que não existe relação entre as duas variáveis e, portanto, elas são estatisticamente independentes.

- A hipótese de pesquisa ($H_a$) declara que as duas variáveis ​estão relacionadas na população, i.e., são estatisticamente dependentes.


# Estatística qui-quadrado

A estatística $\chi^2$ resume as diferenças entre as frequências observadas ($f_o$) e esperadas ($f_e$) em uma tabela de contingência.

$$
\chi^2 = \sum \frac{(f_e - f_o)^2}{f_e}
$$

- Frequências esperadas (*expected frequencies*): valores que seriam esperados se as duas variáveis fossem estatisticamente independentes.
Calculadas para cada par de valores das variáveis.
- Frequências observadas (*observed frequencies*): valores efetivamente observados na tabela de contingência (ou derivados dela).

Para computar a estatística $\chi^2$, é necessário computar as frequências esperadas para cada célula da tabela de contingência. Para isso, basta multiplicar os totais de linha e de coluna para essa célula e dividir o produto resultante pelo número total de casos na tabela (representado por $N$ na equação abaixo).

$$
f_e = \frac{\text{total da coluna} \times {\text{total da linha}}}{N}
$$

Como exemplo, retornemos à tabela de contingência usada em exemplo anterior e reapresentada abaixo.

![link text](https://i.stack.imgur.com/8YuSA.jpg)

As frequências esperadas para cada célula da tabela acima são apresentadas abaixo.

- $f_e$(Male, Dog) = $\frac{51 \times 52}{100} = 26.52$

- $f_e$(Male, Cat) = $\frac{49 \times 52}{100} = 25.48$

- $f_e$(Female, Dog) = $\frac{51 \times 48}{100} = 24.48$

- $f_e$(Female, Cat) = $\frac{49 \times 48}{100} = 23.52$

De posse das frequências observadas, é possível computar o valor da estatística:

$$
\chi^2 = \frac{26.52-42}{26.52} + \frac{25.48-10}{25.48} + \frac{24.48-9}{24.48} + \frac{23.52-39}{23.52} \approx 38.41
$$

In [ ]:
import numpy as np

f_11 = 51*52/100
f_12 = 49*52/100
f_21 = 51*48/100
f_22 = 49*48/100
print(f_11)
print(f_12)
print(f_21)
print(f_22)

print(np.square(f_11 - 42)/f_11)

qui_square = np.square(f_11 - 42)/f_11 + np.square(f_12 - 10)/f_12 + np.square(f_21 - 9)/f_21 + np.square(f_22 - 39)/f_22
print(qui_square)

26.52
25.48
24.48
23.52
9.0358371040724
38.41767476221258


# Distribuição $\chi^2$

A estatística usada no teste $\chi^2$ segue a distribuição hom6onima, a [distribuição $\chi^2$](https://en.wikipedia.org/wiki/Chi-squared_distribution). A figura a seguir ([fonte](https://en.wikipedia.org/wiki/Chi-squared_distribution)) apresenta os gráficos de diversas funções de densidade que seguem a distribuição $\chi^2$. Essa distribuição possui um parâmetro, $k$, a quantidade de graus de liberdade.

![](https://upload.wikimedia.org/wikipedia/commons/thumb/3/35/Chi-square_pdf.svg/1024px-Chi-square_pdf.svg.png)


A biblioteca scipy disponibiliza o pacote de funções [chi2](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2.html), que permite manipular variáveis aleatórias que seguem a distribuição chi-quadrado.

Como exemplo, suponha uma variável aleatória que siga a distribuição qui-quadrado com 9 graus de liberdade. Se quisermos saber que valor dessa variável delimita 95% da distribuição, podemos usar a função ppf, conforme o bloco de código abaixo.

In [ ]:
from scipy.stats import chi2
chi2.ppf(.95, df=9)

inf

# Aplicação do teste

No contexto do teste $\chi^2$, a distribuição $\chi^2$
indica a probabilidade de obter um determinado valor, presumindo que não existe relação (na população) entre as duas variáveis. Ainda no contexto desse teste, o valor do parâmetro $k$ é calculado por meio da expressão

$$
k = (r – 1) \times (c – 1)
$$

onde
- $r$ = quantidade de linhas da tabela, e
- $c$ = quantidade de colunas da tabela.

Seguindo com o exemplo anterior, o parâmetro $k$ pode ser calcular conforme a seguir:

$$
k = (2-1)(2-1) = 1
$$

De posse dos valores da estatística ($\chi^2 = 38.41$) e do parâmetro ($k=1$), podemos computar o valor corresponde de $p$.

Considere que foi definido um nível de significância igual a 5% para o estudo estatístico. Usando a função [scipy.stats.chi2.ppf](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2.html#scipy.stats.chi2), concluímos que o valor $3.84$, conhecido como valor crítico, é o que separa 5% da área da curva à direita. Por outro lado, o valor $\chi^2 = 38.41$ é muito maior do que $3.84$. De fato, o valor obtido para a estatística está relacionado um valor $p$ virtualmente igual a $0$. Então parece muito claro que as variáveis ​​categóricas são **dependentes** na população da qual a amostra de 100 observações foi escolhida aleatoriamente.

In [ ]:
from scipy.stats import chi2
x = 38.41
k=1
print('valor crítico: %.2f' % chi2.ppf(.95,k))
print('p: %f' % (1 - chi2.cdf(x,k)))

valor crítico: 3.84
p: 0.000000


In [ ]:
from scipy.stats import chisquare
chisquare([42, 10, 9, 39],[26.52, 25.48, 24.48, 23.52])

Power_divergenceResult(statistic=38.41767476221258, pvalue=2.3057195936791932e-08)

In [ ]:
from scipy.stats import chisquare
chisquare([25, 25, 25, 25],[26.52, 25.48, 24.48, 23.52])

Power_divergenceResult(statistic=0.20033654487436, pvalue=0.9775349496246623)

# Condições de aplicabilidade

Diferente dos testes estudados até aqui (z-teste, t-teste, ANOVA), o teste $\chi^2$ não requer suposições sobre a distribuição da população da qual a amostra é retirada. Ou seja, o $\chi^2$ é um teste estatístico não-paramétrico.

No entanto, assim como outras técnicas inferenciais, esse teste presume que foi realizada amostragem aleatória.

Além disso, a frequência esperada de cada célula da tabela de contingência deve ser ao menos igual a 5 para que o resultado de aplicação do teste seja confiável.


# Exemplo

O conjunto de dados survey contém dados sobre hábitos de fumo e de atividades físicas de estudantes, dentre outras informações.

- hábitos de tabagismo (`Smoke`)
- nível de prática de exercícios físicos (`Exer`)

Seja usar o teste $\chi^2$ para verificar, nível de significância de 5%, se as variáveis são independentes.

In [ ]:
import pandas as pd
df_survey = pd.read_csv('survey.csv')
df_survey[['Smoke', 'Exer']].head()

Smoke  Exer
0  Never  Some
1  Regul  None
2  Occas  None
3  Never  None
4  Never  Some

O pacote Pandas fornece a função crosstab que pode ser usada para montar a tabela de contingência para a apilcação do teste $\chi^2$. Veja a célula de código a seguir.

In [ ]:
tbl = pd.crosstab(df_survey.Smoke, df_survey.Exer)
tbl

Exer   Freq  None  Some
Smoke                  
Heavy     7     1     3
Never    87    18    84
Occas    12     3     4
Regul     9     1     7

Por inspeção na tabela de contingência acima, percebemos que a variável `Exer` contém três valores, e a variável `Smoke` contém quatro valores.

De posse da tabela de contingência, podemos usar a função scipy.stats.chi2_contingency para aplicar diretamente o teste $\chi^2$.

In [ ]:
from scipy.stats import chi2_contingency
chi2_contingency(tbl)

(5.488545890584232,
 0.48284216946545633,
 6,
 array([[ 5.36016949,  1.0720339 ,  4.56779661],
        [92.09745763, 18.41949153, 78.48305085],
        [ 9.25847458,  1.85169492,  7.88983051],
        [ 8.28389831,  1.65677966,  7.05932203]]))

Ao ser invocada, a função chi2_contingency retorna uma tupla de componentes conforme descrição abaixo:

- o primeiro componente é o valor Qui-quadrado ($5.49$) ;
- o segundo componente corresponde a $p$ ($0.4828$);
- o terceiro componente é a quantidade de graus de liberdade (3);
- o quarto componente corresponde às frequências esperadas, na forma de uma matriz.

Porque o valor de $p = 0.4828$ é maior do que o nível de significância de $\alpha=0.05$, não temos evidência para rejeitar a hipótese nula de que o hábito de fumar seja independente do nível de exercício dos alunos.

No resultado acima, repare que nem todas as frequências esperadas são maiores que 5. Em geral, os resultados do teste $\chi^2$ são confiáveis apenas quando todas as frequências esperadas são maiores ou iguais a 5. Para contornar esse problema, podemos combinar os valores 'None' e 'Some' da variável `Exer` em um único valor, conforme a célula de código abaixo.

In [ ]:
df_survey.replace({'Exer': {'None': 'None_or_Some', 'Some': 'None_or_Some'}}, inplace=True)
df_survey[['Smoke', 'Exer']].head()

Smoke          Exer
0  Never  None_or_Some
1  Regul  None_or_Some
2  Occas  None_or_Some
3  Never  None_or_Some
4  Never  None_or_Some

In [ ]:
tbl = pd.crosstab(df_survey.Smoke, df_survey.Exer)
tbl

Exer   Freq  None_or_Some
Smoke                    
Heavy     7             4
Never    87           102
Occas    12             7
Regul     9             8

In [ ]:
from scipy.stats import chi2_contingency
chi2_contingency(tbl)

(3.2328182226131843, 0.35710308004083174, 3, array([[ 5.36016949,  5.63983051],
        [92.09745763, 96.90254237],
        [ 9.25847458,  9.74152542],
        [ 8.28389831,  8.71610169]]))

Perceba que agora todos os valores das frequências esperadas estão acima de 5, fazendo com que o resultado de aplicação do $\chi^2$ dessa vez seja confiável.